lib/apimatic-core/utilities/union_type_helper.rb

```diff
@@ -0,0 +1,294 @@
+
+      [is_valid, collection_cases]
+    end
+
+    def self.process_array_items(union_types, array_value, is_for_one_of)
+      is_valid = true
+      collection_cases = []
+
+      array_value.each do |item|
+        union_type_cases = make_deep_copies(union_types)
+        matched_count = get_matched_count(item, union_type_cases, is_for_one_of)
+        is_valid = check_item_validity(is_for_one_of, is_valid, matched_count)
+        collection_cases << union_type_cases
+      end
+
+      [is_valid, collection_cases]
+    end
+
+    def self.check_item_validity(is_for_one_of, is_valid, matched_count)
+      if is_valid && is_for_one_of
+        is_valid = matched_count == 1
+      elsif is_valid
+        is_valid = matched_count >= 1
+      end
+      is_valid
+    end
+
+    def self.make_deep_copies(union_types)
+      nested_cases = []
+      union_types.each do |union_type|
+        nested_cases << union_type.dup
+      end
+      nested_cases
+    end
+
+    def self.get_matched_count(value, union_types, is_for_one_of)
+      matched_count = get_valid_cases_count(value, union_types)
+
+      if is_for_one_of && matched_count == 1
+        return matched_count
+      elsif !is_for_one_of && matched_count.positive?
+        return matched_count
+      end
+
+      matched_count = handle_discriminator_cases(value, union_types)
+      matched_count
+    end
+
+    def self.get_valid_cases_count(value, union_types)
+      union_types.count { |union_type| union_type.validate(value).is_valid }
+    end
+
+    def self.handle_discriminator_cases(value, union_types)
+      has_discriminator_cases = union_types.all? do |union_type|
+        union_type.union_type_context.discriminator && union_type.union_type_context.discriminator_value
+      end
+
+      if has_discriminator_cases
+        union_types.each do |union_type|
+          union_type.union_type_context.discriminator = nil
+          union_type.union_type_context.discriminator_value = nil
+        end
+
+        get_valid_cases_count(value, union_types)
+      else
+        0
+      end
+    end
+
+    def self.optional_or_nullable_case?(current_context, inner_contexts)
+      current_context.nullable_or_optional? || inner_contexts.any?(&:nullable_or_optional?)
+    end
+
+    def self.update_nested_flag_for_union_types(nested_union_types)
+      nested_union_types.each do |union_type|
+        union_type.union_type_context.is_nested = true
+      end
+    end
+
+    def self.invalid_array_value?(value)
+      value.nil? || !value.instance_of?(Array)
+    end
+
+    def self.invalid_dict_value?(value)
+      value.nil? || !value.instance_of?(Hash)
+    end
+
+    def self.serialize_value(value, context, collection_cases, union_types)
+      return serialize_array_of_dict_case(value, collection_cases) if
+        context.is_array && context.is_dict && context.is_array_of_dict
+      return serialize_dict_of_array_case(value, collection_cases) if
```

Comment:
Nits, Non-blocker: Add empty lines after each return to have a better readability.

Reply:
done